peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+those of :py:data:`typing.TypeGuard`.
+
+It accepts a single
+argument and can be used as the return type of a function. A function annotated as returning a
+``TypeNarrower`` is called a type narrowing function. Type narrowing functions must return ``bool``
+values, and the type checker should verify that all return paths return
+``bool``. However, ``TypeNarrower`` is not a subtype of ``bool``.
+The type ``Callable[..., TypeNarrower[int]]`` is not assignable to
+``Callable[..., bool]`` or ``Callable[..., TypeGuard[int]]``, and vice versa.
+
+Type narrowing functions must accept at least one positional argument. The type
+narrowing behavior is applied to the first positional argument passed to
+the function. The function may accept additional arguments, but they are
+not affected by type narrowing.
+
+Type narrowing behavior
+-----------------------
+
+To specify the behavior of ``TypeNarrower``, we use the following terminology:
+
+* I = ``TypeNarrower`` input type
+* R = ``TypeNarrower`` return type
+* A = Type of argument passed to type narrowing function (pre-narrowed)
+* NP = Narrowed type (positive; used when ``TypeNarrower`` returned True)
+* NN = Narrowed type (negative; used when ``TypeNarrower`` returned False)
+
+.. code-block:: python
+
+    def narrower(x: I) -> TypeNarrower[R]: ...
+
+    def func1(val: A):
+        if narrower(val):
+            assert_type(val, NP)
+        else:
+            assert_type(val, NN)
+
+The return type ``R`` must be :ref:`consistent with <pep-483-gradual-typing>` ``I``. The type checker should
+emit an error if this condition is not met.
+
+Formally, type *NP* should be narrowed to :math:`A \land R`, the intersection of *A* and *R*, and type *NN* should be narrowed to
+:math:`A \land \neg R`, the intersection of *A* and the complement of *R*.
+In practice, the theoretic types for strict type guards cannot be expressed
+precisely in the Python type system. Type checkers should fall back on
+practical approximations of these types. As a rule of thumb, a type checker
+should use the same type narrowing logic -- and get results that are consistent
+with -- its handling of :py:func:`isinstance`. This guidance allows for changes and
+improvements if the type system is extended in the future.
+
+In particular, if the argument type ``A`` is a union type, ``NP`` may
+be narrowed to those elements of the union that are consistent with ``R``,
+and ``NN`` may be narrowed to those elements of the union that are
+not subtypes of ``R``.
+
+Examples
+--------
+
+Type narrowing is applied in both the positive and negative case::
+
+    from typing import TypeNarrower, assert_type
+
+    def is_str(x: object) -> TypeNarrower[str]:
+        return isinstance(x, str)
+
+    def f(x: str | int) -> None:
+        if is_str(x):
+            assert_type(x, str)
+        else:
+            assert_type(x, int)
+
+Additional type narrowing is applied to the positive case if applicable::
```

Comment:
I don't think "Additional type narrowing is applied" is a clear description of what follows; "additional" suggests it is narrowing that comes _after_ the TypeNarrower, where really it is taking into account type information previously known. I would rather say something like "The final narrowed type may be narrower than R, due to the constraints of its previously-known type:"

Reply:
That's a better wording, will change this.